torch/csrc/autograd/record_function.h

```diff
@@ -281,36 +277,92 @@ TORCH_API void TEST_unsetGlobalSamplingProbability();
+//       invoked by every RecordFunction, in addition to the thread local
+//       callbacks specific to the given thread
+//     - we allow global callbacks to be sampled, by specifying a sampling
+//       probability for each global callback pair, if the start callback is
+//       not picked to run, the corresponding end callback won't be called
+//     - a typical use case for the global callbacks is passive monitoring
+//       in the background (e.g. fleet-wide monitoring), without focusing on
+//       the specific peice of code
+//     - in contrast, thread local callbacks are enabled locally, on demand,
+//       for specific piece of code (range) and are not sampled
+//     - a typical use case for thread local callbacks is profiler and
+//       code execution tracer
+//       - note, that some functionality (e.g. profiler) can automatically
+//         propagate itself across threads by using ThreadLocalState mechanism
+//     - unlike global callbacks, thread local callbacks do not require
+//       RecordFunction to store their state internally and allow an easy,
+//       frequent and thread-safe adding/removing of callbacks for a particular
+//       thread
+//     - global callbacks (pairs) are sampled, this requires RecordFunction
+//       to store the state of their execution (which callbacks were picked and
+//       the 'version' of the global list of the callbacks)
+//     - for performance reasons RecordFunction does not copy the callbacks
+//       into the internal state
+//     - adding/removing global callbacks is not thread safe and should be done
+//       only when no other code is running, e.g. during the initialization
+
 /**
- * pushCallback adds a pair of callbacks to run with RecordFunction:
- *  start, end - the callbacks to run when entering and exiting the scope;
- *    if start callback returns false, end callback won't be executed;
- *  needs_inputs - whether the callbacks need the inputs passed from the observed
- *    function/range; NOTE: passing the inputs incurs an additional overhead;
- *  sampling_prob - whether the callbacks are sampled and the sampling
- *    probability;
- *  scopes - types of scopes to execute the callbacks on (see RecordScope);
- *    passing empty set means the callbacks will be executed for all possible
- *    scope types
- *
- * WARNING: not thread safe, must not overlap with other PyTorch code execution
+ * pushCallback adds a pair of thread local callbacks to run with RecordFunction:
+ *   start, end - the callbacks to run when entering and exiting the scope;
+ *   needs_inputs - whether the callbacks need the inputs passed from the observed
+ *     function/range; NOTE: passing the inputs incurs an additional overhead;
+ *   scopes - types of scopes to execute the callbacks on (see RecordScope);
+ *     passing empty set means the callbacks will be executed for all possible
+ *     scope types
+ *  is_thread_local - whether this pair of callbacks should be set
```

Comment:
Might be missing something, but I actually don't see a `is_thread_local` arg below?

Reply:
right, will upd